sui_programmability/examples/regulated_coin/sources/RegulatedCoin.move

```diff
@@ -0,0 +1,431 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+/// Module representing a common type for regulated coins. Features balance
+/// accessors which can be used to implement a RegulatedCoin interface.
+///
+/// To implement any of the methods, module defining the type for the currency
+/// is expected to implement the main set of methods such as `borrow()`,
+/// `borrow_mut()` and `zero()`.
+///
+/// Each of the methods of this module requires a Witness struct to be sent.
+module RC::RegulatedCoin {
+    use Sui::Balance::{Self, Balance};
+    use Sui::TxContext::{Self, TxContext};
+    use Sui::ID::VersionedID;
+
+    /// The RegulatedCoin struct; holds a common `Balance<T>` which is compatible
+    /// with all the other Coins and methods, as well as the `owner` field, which
+    /// can be used for additional security/regulation implementations.
+    struct RegulatedCoin<phantom T> has key {
+        id: VersionedID,
+        balance: Balance<T>,
+        owner: address
```

Comment:
I don't fully understand the need for the `owner` field. Do you anticipate a case where the Sui owner of the coin is different from the `owner` field of the coin?

Reply:
Yes indeed. And also it opens up a possibility to create lists of banned addresses. It is one of the possible usages of a regulated coin.

Reply:
To implement banned addresses, you can always just check the sender address?
I still don't know when you would want the sender to be different from the owner field.

Reply:
Not quite. I assume that it is possible to send the whole struct to another user so he can withdraw as the sender address differs. By adding this field we give developers an option to prevent a scenario like this.

Reply:
> Not quite. I assume that it is possible to send the whole struct to another user so he can withdraw as the sender address differs. By adding this field we give developers an option to prevent a scenario like this.

I think that is a thing we need to prevent. I think it would be better to prevent the transfer than to add a secondary idea of ownership in the system. (though that doesn't necessarily mean we don't need an owner field, not sure. Just feel that the transfer should be regulated too)

Reply:
But in general what is going on here feels reasonable for the way transfers work today. But I think changes to how restricted transfers might work in the future will change this dramatically.